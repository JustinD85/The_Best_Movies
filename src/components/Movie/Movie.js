import React, { Fragment } from 'react'
import { Link } from 'react-router-dom'
import { postData, getData, deleteData } from '../../api'
import { setShouldPromptLogin, setFavorites, setPopup } from '../../actions'
import { connect } from 'react-redux'
import './Movie.scss'

export const Movie = (
  {
    movie, user, setShouldPromptLogin, setFavorites, isPopup, setPopup
  }
) => {
  const imgUrl = 'https://image.tmdb.org/t/p/w500'
  const checkCanFavorite = async () => {
    await setShouldPromptLogin(!user)
    if (user) {
      const pathToAddFav = '/users/favorites/new'
      const pathToGetFav = `users/${user.id}/favorites`
      const pathToDeleteFav = `${pathToGetFav}/${movie.id}`
      try {
        !movie.isFavorite ?
          await postData(pathToAddFav, movie) :
          await deleteData(pathToDeleteFav, movie)
        const favorites = await getData(pathToGetFav)
        setFavorites(favorites, user.id)
      } catch (error) {
        console.log(error)
      }
    }
  }
  const getImage = () => {
    return <Link to={`/movies/${movie.id}`}  >
      <img
        onClick={() => setPopup(true)}
        src={imgUrl + movie.poster_path}
        className='movie-poster'
        alt='movie poster'></img>
    </Link>
  }
  const getTrailer = () => {
    return <iframe width="560" height="315" title={movie.title}
      src={`https://www.youtube.com/embed/${movie.trailer}`}
      frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
      allowFullScreen></iframe>
  }
  const _class = isPopup ? 'popup' : 'Movie'

  return (
    <div className={_class} >
      {isPopup && <Fragment>
        <Link className='close-popup' onClick={() => setPopup(false)} to='/'>X</Link>
        <h1 className='movie-title'>{movie.title}</h1>
        <hr/>
        <p className='movie-release'>Released: {movie.release_date}</p>
        <hr/>
        {getTrailer()}
        <hr/>
        <p className='movie-description'>{movie.overview}</p>
        <hr/>
        <p className='movie-vote'>Average rating: {movie.vote_average}</p>
      </Fragment>}

      {!isPopup && getImage()}

      <span className='favorite-btn' onClick={() => checkCanFavorite()}>{movie.isFavorite ? '⭐️' : '☆'}</span>
    </div>
  )
}
const mapStateToProps = (state) => ({
  user: state.user,
})

const mapDispatchToProps = (dispatch) => ({
  setFavorites: (favorites, user_id) => dispatch(setFavorites(favorites, user_id)),
  setShouldPromptLogin: (bool) => dispatch(setShouldPromptLogin(bool)),
  setPopup: (bool) => dispatch(setPopup(bool)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie)