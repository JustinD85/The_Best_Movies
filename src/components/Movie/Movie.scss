.Movie {
  margin: 0px 10px;
  color: #ECEBEE;
}

.movie-poster {
  height: 600px;
  border-radius: 15px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  transition: all 0.3s cubic-bezier(.25, .8, .25, 1);

  &:hover {
    height: 602px;
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
}

.popup {
  z-index: 1;
  position: absolute;
  left: 4%;
  top: 10%;
  border-radius: 15px;
  background: rgba(0, 0, 0, 0.97);
  color: #ECEBEE;
  padding: 0px 30px 30px 30px;
  height: 64%;
  width: 84%;
}

.movie-title {
  letter-spacing: 2px;
}

.close-popup {
  top: 0%;
  right: 92%;
  color: #ECEBEE;
  font-weight: 700;
  text-decoration: none;
  border: 2px solid #5a5a5a8a;
  padding: 5px 8px;
  border-radius: 100px;
  margin: 10px;
  position: absolute;
}

hr {
  width: 71%;
  padding: 1px;
  margin: 4px auto;
}

.movie-release,
.movie-description,
.movie-vote {
  font-size: 18px;
  width: 74%;
  margin: auto;
}

.favorite-btn {
  display: block;
  margin: auto;
  position: relative;
  top: -91.5%;
  left: 42%;
  font-size: 34px;
  
  &:hover {
    cursor: pointer;
  }
}