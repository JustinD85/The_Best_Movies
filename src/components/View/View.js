import React, { Component } from 'react'
import { Route } from 'react-router-dom'
import { Profile, Carousel } from '../../containers'
import { Movie } from '../../components'
import { Nav } from '../'
import './View.scss'

class View extends Component {
  
  findClickedCard = (match) => {
    const { movies } = this.props
    const movie = movies.find(movie => parseInt(movie.id) === parseInt(match.params.id))
    if (!movie) return <div> Movie does not exist</div>
    return <Movie key={JSON.stringify(movie)} movie={movie} isPopup={true} />
  }
  
  render() {
    return (
      <div className="Home">
        <Nav />
        <section className='View'>
          <Route path='/' component={Profile} />
          <Route path='/' render={({ location }) => <Carousel location={location} />} />
          <Route path='/movies/:id' render={({ match }) => this.findClickedCard(match)} />
        </section>
      </div>
    )
  }
}

export default View