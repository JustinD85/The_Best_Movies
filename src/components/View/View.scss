.View {
  background: #0f171e;
  width: 80%;
  height: 100vh;
  position: relative;
}

.Home {
  display: flex;
}

.popup {
  height: 54%;
  width: 80%;
  z-index: 1;
  position: absolute;
  left: 4%;
  top: 10%;
  border-radius: 25px;
  background: rgba(0, 0, 0, .5)
}