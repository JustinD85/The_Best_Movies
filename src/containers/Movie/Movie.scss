.Movie {
  width: 30%;
  border: 1px solid white;
  margin: 10px;
}

.movie-title {
  color: #ECEBEE;
}

.movie-poster {
  height: 300px;
}

.favorite-btn {
  display: block;
  margin: auto;
}