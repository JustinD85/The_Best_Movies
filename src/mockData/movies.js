export const mockMovie = {
    "id": "5a7ee86cc3a368189300b429",
    "iso_639_1": "en",
    "iso_3166_1": "US",
    "key": "2mNMvz85NG8",
    "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
    "site": "YouTube",
    "size": 720,
    "type": "Teaser",
}

export const mockMoviesResponse = {
    "id": 450465,
    "results": [
        {
            "id": "5a7ee86cc3a368189300b429",
            "iso_639_1": "en",
            "iso_3166_1": "US",
            "key": "2mNMvz85NG8",
            "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
            "site": "YouTube",
            "size": 720,
            "type": "Teaser",
        }
    ]
}

export const mockMoviesArray = [
    {
        "id": "5a7ee86cc3a368189300b429",
        "iso_639_1": "en",
        "iso_3166_1": "US",
        "key": "2mNMvz85NG8",
        "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
        "site": "YouTube",
        "size": 720,
        "type": "Teaser",
    }
]