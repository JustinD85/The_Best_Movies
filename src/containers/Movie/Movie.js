import React from 'react'
import './Movie.scss'

const Movie = ({movie}) => {
  return (
    <div className='Movie'>
      <h1 className='movie-title'>{movie.title}</h1>
      <p className='movie-title'>{movie.release_date}</p>
    </div>
  )
}

export default Movie;