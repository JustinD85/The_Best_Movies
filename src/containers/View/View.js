import React from 'react'
import Carousel from '../../components/Carousel/Carousel'
import { Route } from 'react-router-dom'
import './View.scss'

export default () => {
  return (
    <div className='View'>
      <Route path='/' component={Carousel} />
    </div>)
}