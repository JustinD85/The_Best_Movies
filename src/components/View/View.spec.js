import React from 'react'
import { shallow } from 'enzyme'
import { Route } from 'react-router-dom'
import View from './View'
import Movie from '../Movie/Movie'
import { mockMoviesArray, mockMovie } from '../../mockData/movies'

describe('View', () => {
  it('should match the snapshot', () => {
    let wrapper = shallow(<View movies={mockMoviesArray} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 3 routes', () => {
    let wrapper = shallow(<View movies={mockMoviesArray} />)
    expect(wrapper.find(Route).length).toEqual(3)
  })

  it('should invoke findClickedCard ', () => {
    let wrapper = shallow(<View movies={mockMoviesArray} />)
    let result = wrapper.instance().findClickedCard({ params: { id: "91" } })
    let expected = <Movie key={JSON.stringify(mockMovie)} movie={mockMovie} isPopup={true} />
    expect(result).toEqual(expected)
  })

  it('should invoke findClickedCard ', () => {
    let wrapper = shallow(<View movies={mockMoviesArray} />)
    let result = wrapper.instance().findClickedCard({ params: { id: "92" } })
    let expected = <div> Movie does not exist</div>
    expect(result).toEqual(expected)
  })
})