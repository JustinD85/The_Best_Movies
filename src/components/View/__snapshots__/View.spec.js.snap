// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View should match the snapshot 1`] = `
<div
  className="Home"
>
  <Nav />
  <section
    className="View"
  >
    <Route
      component={[Function]}
      path="/"
    />
    <Route
      path="/"
      render={[Function]}
    />
    <Route
      path="/movies/:id"
      render={[Function]}
    />
  </section>
</div>
`;
